{
  "comments": [
    {
      "unresolved": false,
      "key": {
        "uuid": "bc3e767a_415ab454",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 92,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "This was actually not possible. `is_reapable \u003d\u003d False` \u003d\u003e `expiration_ts \u003d\u003d None` and the `if` above is fired.",
      "range": {
        "startLine": 90,
        "startChar": 0,
        "endLine": 92,
        "endChar": 57
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "4b28a99f_27ee7125",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 101,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "Skipping _expire_task_tx just because there\u0027s something with clocks across machines (which I\u0027m 70% sure is the reason for this condition) will not be great when _expire_task_tx is not being called repeatedly from a cron. If _expire_task_tx is called, it should do the change.\n\nAlso I could not find recent occurrence of this condition in logs.",
      "range": {
        "startLine": 101,
        "startChar": 4,
        "endLine": 101,
        "endChar": 28
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "e83eda3a_b521bdc6",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 113,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "This is super weird and I\u0027m 80% was the cause of the bug mentioned below. I replaced it with cleaner variant without \"small race condition\".",
      "range": {
        "startLine": 111,
        "startChar": 2,
        "endLine": 113,
        "endChar": 27
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "11bb8b8c_4295464c",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 113,
      "author": {
        "id": 1547372
      },
      "writtenOn": "2023-03-02T16:55:55Z",
      "side": 0,
      "message": "Just making sure I understand what\u0027s happening here - the race condition is fixed by using `task_to_run.task_to_run_key_slice_index` instead of `result_summary.current_task_slice` to create the `slice_index`\n\nBefore the `result_summary.current_task_slice` may not have been up to date with the `to_run_key` - IE the actual `TaskToRunShardXXX` we are expiring.",
      "parentUuid": "e83eda3a_b521bdc6",
      "range": {
        "startLine": 111,
        "startChar": 2,
        "endLine": 113,
        "endChar": 27
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "d1c18d83_2429d4fc",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 113,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T18:14:49Z",
      "side": 0,
      "message": "Yes.\n\nAlso I strongly suspect `current_task_slice` is often wrong on chromeos-swarming when using external scheduler, since `_ensure_active_slice` called there doesn\u0027t update `current_task_slice` for some reason (not sure if intentional or a bug).\n\nI also realized `_expire_task_tx` was doing something extremely weird in this case. It was picking the next slice based on (potentially wrong) `result_summary.current_task_slice`, totally ignoring `task_to_run.task_to_run_key_slice_index`. It means it could have been reenqueuing the exact same slice it just expired. I just hope it is not something chromeos-swarming actually accidentally relies on.\n\nI also don\u0027t understand why this function consults `capacity` at all when using External Scheduler. `capacity` is populated based on Swarming native scheduler state.\n\nThis is all full of bugs ðŸ˜­ I\u0027ll deploy this and hope it doesn\u0027t blow up.",
      "parentUuid": "11bb8b8c_4295464c",
      "range": {
        "startLine": 111,
        "startChar": 2,
        "endLine": 113,
        "endChar": 27
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "c0102b80_1cf005f9",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 143,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "This is not possible anymore, retries were removed.",
      "range": {
        "startLine": 143,
        "startChar": 4,
        "endLine": 143,
        "endChar": 32
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "68fa3bc1_1eb14b5f",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 157,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "No it isn\u0027t in this code path. Some slices can be skipped immediately without waiting if there are no bots that can execute them.",
      "range": {
        "startLine": 157,
        "startChar": 28,
        "endLine": 157,
        "endChar": 66
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    },
    {
      "unresolved": false,
      "key": {
        "uuid": "703d7d46_ecf9cc76",
        "filename": "appengine/swarming/server/task_scheduler.py",
        "patchSetId": 2
      },
      "lineNbr": 248,
      "author": {
        "id": 1002539
      },
      "writtenOn": "2023-03-02T04:28:33Z",
      "side": 0,
      "message": "This is a waste of time, we just stored it.",
      "range": {
        "startLine": 248,
        "startChar": 44,
        "endLine": 248,
        "endChar": 60
      },
      "revId": "d5da1b4f0b60e1eea227849b808bb00398162003",
      "serverId": "3ce6091f-6c88-37e8-8c75-72f92ae8dfba"
    }
  ]
}